django/db/models/deletion.py

```diff
@@ -219,7 +220,12 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
                     if self.can_fast_delete(sub_objs, from_field=field):
                         self.fast_deletes.append(sub_objs)
                     elif sub_objs:
-                        field.remote_field.on_delete(self, field, sub_objs, self.using)
+                        try:
+                            field.remote_field.on_delete(self, field, sub_objs, self.using)
+                        except ProtectedError as e:
+                            protected_objects += e.protected_objects
+            if protected_objects:
+                raise ProtectedError('', protected_objects)
```

Comment:
I think we should try to preserve the original exception in the more common case where only one relationship is protected. In other words, add this test in a separate commit before this fix:
```diff
diff --git a/tests/delete/tests.py b/tests/delete/tests.py
index aa77fe5..c7ab78b 100644
--- a/tests/delete/tests.py
+++ b/tests/delete/tests.py
@@ -60,7 +60,11 @@ class OnDeleteTests(TestCase):
 
     def test_protect(self):
         a = create_a('protect')
-        with self.assertRaises(IntegrityError):
+        msg = (
+            "Cannot delete some instances of model 'R' because they are "
+            "referenced through a protected foreign key: 'A.protect'"
+        )
+        with self.assertRaisesMessage(IntegrityError, msg):
             a.protect.delete()
 
     def test_protect_multiple(self):
```
Then for the case where there are multiple relations affected, let's use a more descriptive message at Collin suggested, perhaps something like ""Cannot delete some instances of model '%s' because they are referenced through a protected foreign keys: 'A', 'B'" would also be useful here.